GoogleForCreators/web-stories-wp
# Dashboard: Add loading spinner to dashboard views

## Context

Adding a loading animation to show the user that content is loading.

Loading spinner PR: #6681

## Summary

Adds the new loading spinner to the dashboard
* add loading spinner to 'my stories' views
    * show big spinner when filters, sort, and search are changed
* add loading spinner to 'templates' views

## Relevant Technical Choices

The tricky parts:
1. we want to show a smaller loading spinner when the 'infinite scroll' is triggered, but we want to hide the story content (and infinite scroll) when refetch is triggered from a change to a filter, search, or sort.
2. we want to display the table header in the story list view when loading
3. we can hide the entire grid view when loading

First: I added a ref to track when we should hide the stories: `showStoriesWhileLoading`. This is created in `useStoryView` and shared by the grid view and list view. Changing this state will not re-render the app since it is a ref.

Second: This ref is updated in the `useStoryView` hook for when stories should be displayed while loading (i.e. infinite scrolling)

Third: After each fetch is completed, the ref is reset so that it is ready for any subsequent

Now, using `isLoading` and `showStoriesWhileLoading` the views can figure out whether they need to display their stories or not. If y'all see a better way to do this, lmk! Feels overly complicated.

## To-do

none

## User-facing changes

|||
|--|--|
|<img src="https://user-images.githubusercontent.com/22185279/110986165-8f3df700-832a-11eb-8fe4-2e3413c3e8a4.gif" width="800px" />|![list-view](https://user-images.githubusercontent.com/22185279/110985746-0b840a80-832a-11eb-8019-8c71df659706.gif)|
|![templates](https://user-images.githubusercontent.com/22185279/110986284-b5fc2d80-832a-11eb-84d9-4c8f804a31f1.gif)|![rtl](https://user-images.githubusercontent.com/22185279/110988543-eabdb400-832d-11eb-9927-d437c2a47124.gif)|

## Testing Instructions

1. Trigger loads on the dashboard by changing the search, sort, and filters. The big spinner and the little spinner (at the bottom of the page) should never show at the same time.
2. Check to see that the text on the page makes sense. It may not make sense to show `No More Stories` at the bottom of the page if the stories are being re-fetched
3. Looking at templates should show you a big spinner
4. Should all be readable by screen reader.

### QA






This PR can be tested by following these steps:

1. Trigger loads on the dashboard by changing the search, sort, and filters. The big spinner and the little spinner (at the bottom of the page) should never show at the same time.
2. Check to see that the text on the page makes sense. It may not make sense to show `No More Stories` at the bottom of the page if the stories are being re-fetched
3. Looking at templates should show you a big spinner

### UAT






## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist




---



Fixes #6622
